packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts

```diff
@@ -44,70 +44,56 @@ export class RemoteMessageProcessor {
-
-			for (let ungroupedMessage2 of this.opGroupingManager.ungroupOp(message)) {
-				// unpack and unchunk the ungrouped message in place
-				unpackRuntimeMessage(ungroupedMessage2);
-				const chunkProcessingResult =
-					this.opSplitter.processRemoteMessage(ungroupedMessage2);
-				ungroupedMessage2 = chunkProcessingResult.message;
-
-				// if the splitter is still rebuilding the original message, there is no need to continue processing
-				if (chunkProcessingResult.state === "Accepted") {
-					continue;
-				}
-
-				// If the message is not chunked there is no need to continue processing
-				if (chunkProcessingResult.state !== "Processed") {
-					result.push(
-						ungroupedMessage2 as InboundSequencedContainerRuntimeMessageOrSystemMessage,
-					);
-					continue;
-				}
-
-				// Ungroup before and after decompression for back-compat (cleanup tracked by AB#4371)
-				for (const ungroupedMessageAfterChunking of this.opGroupingManager.ungroupOp(
-					ungroupedMessage2,
-				)) {
-					const decompressionAfterChunking = this.opDecompressor.processMessage(
-						ungroupedMessageAfterChunking,
-					);
-
-					for (const ungroupedMessageAfterChunking2 of this.opGroupingManager.ungroupOp(
-						decompressionAfterChunking.message,
-					)) {
-						if (decompressionAfterChunking.state === "Skipped") {
-							// After chunking, if the original message was not compressed,
-							// there is no need to continue processing
-							result.push(
-								ungroupedMessageAfterChunking2 as InboundSequencedContainerRuntimeMessageOrSystemMessage,
-							);
-							continue;
-						}
-
-						// The message needs to be unpacked after chunking + decompression
-						unpack(ungroupedMessageAfterChunking2);
-						result.push(ungroupedMessageAfterChunking2);
-					}
-				}
+		let message = remoteMessageCopy;
+		ensureContentsDeserialized(message);
+
+		if (isChunkedMessage(message)) {
+			const chunkProcessingResult = this.opSplitter.processChunk(message);
+			// Only continue further if current chunk is the final chunk
+			if (!chunkProcessingResult.isFinalChunk) {
+				return [];
 			}
+			// This message will always be compressed
+			message = chunkProcessingResult.message;
 		}
 
-		return result;
+		if (this.opDecompressor.isCompressedMessage(message)) {
+			this.opDecompressor.decompressAndStore(message);
+		}
+
+		if (this.opDecompressor.currentlyUnrolling) {
+			message = this.opDecompressor.unroll(message);
+			// Need to unpack after unrolling if not a groupedBatch
+			if (!isGroupedBatch(message)) {
+				unpack(message);
+			}
+		}
+
+		if (isGroupedBatch(message)) {
+			return this.opGroupingManager.ungroupOp(message).map((el) => {
+				// We always need to unpack an ungrouped message
+				unpack(el);
+				return el as InboundSequencedContainerRuntimeMessageOrSystemMessage;
+			});
+		}
+
+		// Do a final unpack of runtime messages in case the message was not grouped, compressed, or chunked
```

Comment:
I think this PR is a great opportunity to document this with some sort of (state transition|flow) diagram (something like https://github.com/microsoft/FluidFramework/pull/17273/files) because even thought it's cleaner and more straight-forward, there are still a lot of possible states here.

Reply:
To clarify, I don't mean that it needs to be part of the PR, I mean that it's a great opportunity to do it in the near future while the refactoring is still fresh in your mind. Doing it later can be as daunting as someone coming in new to this area of the code 😃